src/Features/CSharp/Portable/SignatureHelp/AbstractCSharpSignatureHelpProvider.cs

```diff
@@ -72,5 +100,308 @@ protected IList<TaggedText> GetAwaitableUsage(IMethodSymbol method, SemanticMode
+        /// If it was bound, we'll find which parameter to highlight.
+        /// </summary>
+        protected static (ISymbol symbol, int parameterIndex) GuessCurrentSymbolAndParameter(
+            SeparatedSyntaxList<ArgumentSyntax> arguments, ImmutableArray<IMethodSymbol> methodGroup, int position,
+            SemanticModel semanticModel, ISemanticFactsService semanticFactsService, CancellationToken cancellationToken)
+        {
+            if (arguments.Count != 0)
+            {
+                foreach (var method in methodGroup)
+                {
+                    if (FindParameterIndexIfCompatibleMethod(arguments, method, position, semanticModel, semanticFactsService, out var parameterIndex))
+                    {
+                        return (method, parameterIndex);
+                    }
+                }
+            }
+
+            // Note: Providing no recommendation if no arguments allows the model to keep the last implicit choice
+            return (null, -1);
+        }
+
+        /// <summary>
+        /// If an argument name does not correspond to a parameter name, this method is unacceptable.
+        /// We'll filter it out, so it will not be displayed as a SignatureHelp candidate.
+        /// </summary>
+        protected static bool IsUnacceptable(SeparatedSyntaxList<ArgumentSyntax> arguments, IMethodSymbol method)
+        {
+            if (arguments.Count == 0)
+            {
+                return false;
+            }
+
+            var parameterNames = method.Parameters.SelectAsArray(p => p.Name);
+            foreach (var argument in arguments)
+            {
+                if (HasName(argument, out var name))
+                {
+                    if (!parameterNames.Contains(name))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Simulates overload resolution with the arguments provided so far and determines if you might be calling this overload.
+        /// Returns true if an overload is acceptable. In that case, we output the parameter should be highlighted given the cursor's
+        /// position in the partial invocation.
+        /// </summary>
+        protected static bool FindParameterIndexIfCompatibleMethod(SeparatedSyntaxList<ArgumentSyntax> arguments, IMethodSymbol method, int position,
+            SemanticModel semanticModel, ISemanticFactsService semanticFactsService, out int foundParameterIndex)
+        {
+            var argumentCount = arguments.Count;
+            var parameters = method.Parameters;
+            var parameterCount = parameters.Length;
+            var argToParamMap = ArrayBuilder<int>.GetInstance(argumentCount, -1);
+
+            var result = IsCompatibleMethod(out foundParameterIndex);
+            argToParamMap.Free();
```

Comment:
1. why does the argToParamMap need to live outside of ISCompatibleMethod?  Can it just be moved inside?
2. i find stff like this really hard to reason about because the code looks like it's not mutating/capturing that local.  I'd far prefer static local functions to make it clear since you have to pass the values along.